### PR TITLE
ARROW-5569: [C++] Avro import

### PR DESCRIPTION
**Pull request by emkornfield**
This PR just copies code from avro C++ api/impl/test directories with the following modifications:
1.  Puts all files in one directory
2.  Renames all files to snake_case instead of CamelCase.
3.  Run a make format.

For now files are excluded from linting.

Getting the code to actually build is covered by a separate JIRA and will have a separate PR.  I'm doing this first so subsequent PRs can show more accurately the actual diffs required.

### CODE REVIEW

**Comment by wesm** [comment]
Nice, thanks for taking up this project. I'm guessing we'll be deleting a bunch of code in follow up PRs

**Comment by emkornfield** [comment]
Yes the next PR should have a lot of deletion.

**Comment by emkornfield** [comment]
I believe CI failure was unrelated.

File: cpp/src/arrow/dataset/avro/buffer/buffer.h [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+#include <sys/uio.h>
+#endif
+#include <vector>
+
+#include "../Config.hh"
+#include "detail/BufferDetail.hh"
+#include "detail/BufferDetailIterator.hh"
+
+/**
+ * \file Buffer.hh
```
Comment (kou) [inline cpp/src/arrow/dataset/avro/buffer/buffer.h]: This may cause a CI failure:
https://travis-ci.org/apache/arrow/jobs/546330355#L4009

```text
/home/travis/build/apache/arrow/cpp/src/arrow/dataset/avro/buffer/buffer.h:32: error: the name `Buffer.hh' supplied as the second argument in the \file statement is not an input file (warning treated as error, aborting now)
```

**Comment by emkornfield** [comment]
Nice catch @kou I don't know why I missed that.

**Comment by pitrou** [comment]
Can you explain why we're doing this? This is basically forking a 20KLOC codebase?

**Comment by emkornfield** [comment]
@pitrou the intent was to start from known working code and pair it down/ adapt it to our style standards.  Doing the initial import was meant to make reviewers life easier so it would be obvious what changes were being made from the original code.  At least some code also duplicates existing libraries and will be deleted as part of the adaptation.

The fork in general is because we feel we will ultimately do more advanced things for avro like code generation that would be difficult to merge upstream (and  avoid slow upstream release cycles).

If you prefer I can attempt to do the adaptation as part of this PR or if you have a different preferred strategy I would probably be OK with another approach

**Review by wesm — APPROVED** [review]
+1 -- I'm in favor of this approach. I suspect the codebase will need to be significantly modified to become more "Arrow-native". What do you think about working on Avro in a branch versus in master (either is OK with me)?

**Comment by pitrou** [comment]
Forking sounds like a can of worms to me. How do ensure we remain compatible with upstream? How do we backport fixes and improvements?

I'm also skeptical that this has to be in the Arrow main repo. It seems we're already exploding under the constraints of wrangling together multiple languages and components. Every release we have to firefight building and packaging issues that appear because of various changes (internal or external).

**Comment by emkornfield** [comment]
@pitrou We take ownership of compatibility upstream, luckily the Avro format has been around for a while and fairly stable.  I agree it is a risk. (Same applies to back-ports/fixes).

I think having this in the Arrow repo reduces potential breakages (one less external dependencies), I believe this is why we have Parquet in the main repo also.

**Comment by pitrou** [comment]
I think we are underestimating the maintenance burden here. Five years down the road, how reconcilable will we be with upstream? Probably not at all.

**Comment by wesm** [comment]
The thesis would be that a high quality Arrow-friendly Avro implementation will attract users and developers into the project alike. The reason we are here discussing this at all is that Google relies on Avro for BigQuery platform support, so this is a possible hook to have Google invested in maintenance and support of this part of the project.

Note that Avro support in general is a mess -- I'm aware of at least 4 Python libraries (avro, fastavro, uavro, cyavro) for us, all of which have different levels of features and serialization performance. I think that taking ownership of the code is the least-bad thing we can do given that the Apache Avro community has been stagnant in recent years, though it's improving lately. 

Bottom line is that if the goal is to deliver first-rate Avro support to Arrow users I think this approach is the best option compared with doing work upstream in Apache Avro (where we lack committers and PMC members) and trying to get the Avro community to make releases that are compatible with our much-faster release timelines

**Comment by wesm** [comment]
See C++ commit history in Apache Avro https://github.com/apache/avro/commits/master/lang/c%2B%2B

This looks to me like a project in maintenance-only mode

**Comment by wesm** [comment]
Another perspective: this codebase is almost 10 years old, and primarily has a single developer:

```
$ git shortlog -sn lang/c++/
    60  Thiruvalluvan M. G
    36  Thiruvalluvan M G
    11  Scott Banachowski
     7  Ryan Blue
     6  Douglass Cutting
     5  Doug Cutting
     4  Victor Mota
     4  da77a
     3  Niels Basjes
     3  William Matthews
     2  Aniket Mokashi
     1  Daniel Kulp
     1  Eric Rosenquist
     1  JamesAdams800
     1  Karl Wallner
     1  Marcus Kempe
     1  Marius Van Niekerk
     1  Purushotham Nayak
     1  Romain Geissler
     1  Sandeep Joshi
     1  Thomas White
     1  Tom White
     1  Zoltan Farkas
     1  Zoltan Ivanfi
     1  jmccl
     1  skarlsson
```

This does not look to me like software that is at risk of becoming burdensome due to upstream patches. 

Not only that, but it may require significant changes in order to have better "Arrow affinity".

**Comment by pitrou** [comment]
Well at least I think someone should contact them officially to inform them and gather their feedback (unless this was already done?). Otherwise it will be interpreted as a hostile fork.

**Comment by wesm** [comment]
OK. I can write to the Avro dev@ mailing list

**Comment by wesm** [comment]
See mailing list thread, also cross-posted to Arrow

https://lists.apache.org/thread.html/54625eaae097a89c384de1378b986712fac837a5a5f26bdddb4de7ad@%3Cdev.avro.apache.org%3E

**Comment by wesm** [comment]
No word on the Avro mailing list yet. I'll ping again on Monday

**Comment by emkornfield** [comment]
@pitrou @wesm does waiting until 3 days after the blog post for 0.14.0 release and if we here nothing merging this sound OK?

**Comment by pitrou** [comment]
By the way: is it better to fork and heavily modify, or reimplement from scratch?
Is Avro like Python pickle, representing collections of heterogenous data?
Do we only care about decoding Avro?

**Comment by emkornfield** [comment]
> is it better to fork and heavily modify, or reimplement from scratch?
Good question.  I think both have merits.  I thought this approach would give better traceability to understand where, if any  bugs might have crept in.  If you prefer as part of this PR I can do the first round of edits as a separate commit to validate the assumption that having this code as a starting point makes it easier to review.

> Is Avro like Python pickle, representing collections of heterogenous data?
Is is more like parquet or arrow in that it has a predefined schema for every file

> Do we only care about decoding Avro?
Decoding is the priority but I think encoding might be something we want at some point.  
Encode might also be useful for testing but we run into the chicken or egg problem.

**Comment by pitrou** [comment]
>  Encode might also be useful for testing but we run into the chicken or egg problem.

Perhaps using hard-coded binary test vectors?

**Comment by emkornfield** [comment]
> Perhaps using hard-coded binary test vectors?

That will certainly be required. I think we take both approaches with parquet

**Comment by pitrou** [comment]
For my enlightenment, could you give an example of what an Avro schema for input data would be, and how it would be organized? (columnar? row-wise?)

**Comment by emkornfield** [comment]
> For my enlightenment, could you give an example of what an Avro schema for input data would be, and how it would be organized? (columnar? row-wise?)

[Avro](http://avro.apache.org/docs/current/spec.html) is row oriented.  It also has a file format which divides rows into blocks.  The main difference as far as I can tell with Arrow's schema is null isn't modelled as a first class attribute, instead nullable columns are unions between a "null" type  and concrete type (e.g. int64).  The second difference is slightly different modeling and a difference set of logical types.)  Avro also seems to support recursive schemas which I don't think I will support out of the box.

**Comment by emkornfield** [comment]
@pitrou I pushed a second commit giving a flavor the transformation that needs to happen to the code and also deleted files were we already have functionality in the code base.  Let me know if this looks like it will be easier to review, or if you would prefer I go the other way and only submit parts of the code after they are cleaned up from scratch.

**Comment by wesm** [comment]
I'm satisfied with merging this a few days after the release announcement if no conversation happens. I'm going to put together the blog post here in the next few hours

**Comment by emkornfield** [comment]
@wesm @pitrou do you want to merge this or would you prefer I hold close it and re-open once I have something compiling?

**Comment by pitrou** [comment]
Personally, if we don't plan to have some system for tracking upstream changes, I'd rather see something that compiles and that's cleaned up of unused stuff.

**Comment by emkornfield** [comment]
SGTM, closing for now.

**Comment by wesm** [comment]
@emkornfield would you like me to create a staging branch for Avro work on apache/arrow? When we can merge PRs into that branch

**Comment by emkornfield** [comment]
@wesm its really up to the reviewers, I'm happy to create a lot of smaller PRs that are fully functionality directly against master.
